index.d.ts

```diff
@@ -1,288 +1,559 @@
-    props: TableOptions<D>
-  ): TableOptions<D> & { rows: Row<D>[] }
+  export interface UseRowSelectInstanceProps<D extends object> {
+    toggleRowSelected: (rowPath: IdType<D>, set?: boolean) => void
+    toggleRowSelectedAll: (set?: boolean) => void
+    getToggleAllRowsSelectedProps: (props?: object) => object
+    isAllRowsSelected: boolean
+  }
+
+  export interface UseRowSelectRowProps<D extends object> {
+    isSelected: boolean
+    toggleRowSelected: (set?: boolean) => void
+  }
+  /* #endregion */
 
-  export function usePagination<D = {}>(
-    props: TableOptions<D>
-  ): UsePaginationValues<D>
+  /* #region useRowState */
+  export function useRowState<D extends object = {}>(hooks: Hooks<D>): void
 
-  export function useExpanded<D = {}>(
-    props: TableOptions<D>
-  ): TableOptions<D> & {
-    toggleExpandedByPath: () => any
-    expandedDepth: []
-    rows: Row<D>[]
+  export type UseRowStateOptions<D extends object> = Partial<{
+    initialRowStateAccessor: (row: Row<D>) => object
+  }>
+
+  export interface UseRowStateState<D extends object> {
+    rowState: Partial<{
+      cellState: UseRowStateLocalState<D>
+      rowState: UseRowStateLocalState<D>
+    }>
+  }
+
+  export interface UseRowStateInstanceProps<D extends object> {
+    setRowState: (rowPath: string[], updater: UseRowUpdater) => void
+    setCellState: (
+      rowPath: string[],
+      columnID: IdType<D>,
+      updater: UseRowUpdater
+    ) => void
+  }
+
+  export interface UseRowStateRowProps<D extends object> {
+    state: unknown
+    setState: (updater: UseRowUpdater) => void
+  }
+  export interface UseRowStateCellProps<D extends object> {
+    state: UseRowStateLocalState<D>
+    setState: (updater: UseRowUpdater) => void
+  }
+
+  type UseRowUpdater<T = unknown> = T | ((prev: T) => T)
+  type UseRowStateLocalState<D extends object, T = unknown> = Record<
+    IdType<D>,
+    T
+  >
+  /* #endregion */
+
+  /* #region useSortBy */
+  export function useSortBy<D extends object = {}>(hooks: Hooks<D>): void
+
+  export type UseSortByOptions<D extends object> = Partial<{
+    manualSorting: boolean
+    disableSorting: boolean
+    disableMultiSort: boolean
+    isMultiSortEvent: (e: MouseEvent) => boolean
+    maxMultiSortColCount: number
+    disableSortRemove: boolean
+    disabledMultiRemove: boolean
+    orderByFn: (
+      rows: Row<D>[],
+      sortFns: SortByFn[],
+      directions: boolean[]
+    ) => Row<D>[]
+    sortTypes: Record<string, SortByFn<Row<D>>>
+  }>
+
+  export interface UseSortByHooks<D extends object> {
+    getSortByToggleProps: ((
+      column: Column<D>,
+      instance: TableInstance<D>
+    ) => object)[]
+  }
+
+  export interface UseSortByState<D extends object> {
+    sortBy: SortingRule<D>[]
+  }
+
+  export type UseSortByColumnOptions<D extends object> = Partial<{
+    disableSorting: boolean
+    sortDescFirst: boolean
+    sortInverted: boolean
+    sortType: string | Function // FIXME
```

Comment:
You can use something like this https://github.com/tannerlinsley/react-table/pull/1550/files#diff-b52768974e6bc0faccb7d4b75b162c99R75

```suggestion
    sortType: string | ((rowA: Row<D>, rowB: Row<D>, columnId: IdType<D>) => number)
```